fastestimator/pipeline.py

```diff
@@ -13,442 +13,132 @@
+                warnings.warn("num_process will only be used for built-in dataset")
         else:
-            per_epoch.append(op)
-    return per_epoch
+            raise ValueError("Unsupported dataset type for {}".format(mode))
 
-
-class BasePipeline:
     def get_modes(self) -> Set[str]:
-        raise NotImplementedError()
-
-    def get_batch_size(self, mode: str, epoch: int) -> int:
-        raise NotImplementedError()
-
-    def get_signature_epochs(self, mode: str) -> List[int]:
-        raise NotImplementedError()
-
-    def get_all_output_keys(self) -> Set[str]:
-        raise NotImplementedError()
-
-    def get_num_examples(self, mode: str, epoch: int) -> int:
-        raise NotImplementedError()
-
-    def get_num_steps(self, mode: str, epoch: int) -> int:
-        return self.get_num_examples(mode=mode, epoch=epoch) // self.get_batch_size(mode=mode, epoch=0)
-
-    def get_iterator(self, mode: str, epoch: int = 0, dataset: Optional[Dataset] = None) -> Iterator:
-        raise NotImplementedError()
+        return set(mode for mode in self.data.keys())
 
     def benchmark(self, mode: str = "train", num_steps: int = 1000, log_interval: int = 100, epoch: int = 0):
-        itr = self.get_iterator(mode=mode, epoch=epoch)
+        loader = self.get_loader(mode=mode, epoch=epoch)
         start = time.perf_counter()
-        for idx in range(num_steps + 1):
-            _ = next(itr)
+        for idx, _ in enumerate(loader):
             if idx % log_interval == 0 and idx > 0:
                 duration = time.perf_counter() - start
-                batch_size = self.get_batch_size(mode=mode, epoch=epoch)
-                examples_per_sec = log_interval * batch_size / duration
-                print("FastEstimator: Step: {}, Epoch: {}, Batch Size: {}, Examples/sec: {}".format(
-                    idx, epoch, batch_size, examples_per_sec))
+                iters_per_sec = log_interval / duration
+                print("FastEstimator: Step: {}, Epoch: {}, Steps/sec: {}".format(idx, epoch, iters_per_sec))
                 start = time.perf_counter()
-
-    def show_results(self, mode: str = "train", num_steps: int = 1, epoch: int = 0):
-        data = []
-        itr = self.get_iterator(mode=mode, epoch=epoch)
-        for _ in range(num_steps):
-            data.append(next(itr))
+            if idx == num_steps:
+                break
+
+    def get_loader(self, mode: str, epoch: int = 0):
+        data = self.data[mode]
+        if isinstance(data, Scheduler):
+            data = data.get_current_value(epoch)
+        if isinstance(data, Dataset):
+            op_dataset = OpDataset(data, get_current_ops(self.ops, mode, epoch), mode)
+            batch_size = self.batch_size
+            if isinstance(batch_size, Scheduler):
+                batch_size = batch_size.get_current_value(epoch)
+            data = DataLoader(op_dataset, batch_size=batch_size, shuffle=mode == "train", num_workers=self.num_process)
```

Comment:
drop_last and shuffle should be user overridable, worker_init_fn should be lambda _: np.random.seed()

Reply:
Keeping the `Pipeline` API simple is our primary goal here.  
If user wants more complicated stuff, they are sophisticated enough create a pytorch dataloader themselves and use it in `Pipeline`

Reply:
worker_init_fn needs to be set regardless. 

having drop_last always on and shuffle eval always false means that some eval data will always be ignored

Reply:
I think the default for dataloader is drop last = False, and we are using the default here.    no data will be lost.

Reply:
ah, ok. still need the worker_init_fn though

Reply:
trust me, more than 99% of user won't use that argument.  let's leave it for now.

Reply:
no, not an input argument. It needs to be set to lambda _: np.random.seed()

Reply:
even before this pr I don't think we were letting the user choose that one manually

Reply:
ok, added the worker_init in DataLoader call